### ISSUE TITLE
Panic when calculating median on an array

### ISSUE THREAD

**Issue by marcalj** [liborty/medians#6]
```rust
let mut grades_in_points = [67, 67, 67, 67];

let median_grade = medians::medianu64(&mut grades_in_points)?;
```

Output:

```text
attempt to subtract with overflow
```

```text
attempt to subtract with overflow
stack backtrace:
   0:     0x560b5ef7b962 - std::backtrace_rs::backtrace::libunwind::trace::he4ee80166a02c846
                               at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/../../backtrace/src/backtrace/libunwind.rs:105:5
   1:     0x560b5ef7b962 - std::backtrace_rs::backtrace::trace_unsynchronized::h476faccf57e88641
                               at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x560b5ef7b962 - std::sys_common::backtrace::_print_fmt::h430c922a77e7a59c
                               at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/sys_common/backtrace.rs:68:5
   3:     0x560b5ef7b962 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hffecb437d922f988
                               at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x560b5efa284c - core::fmt::rt::Argument::fmt::hf3df69369399bfa9
                               at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/fmt/rt.rs:142:9
   5:     0x560b5efa284c - core::fmt::write::hd9a8d7d029f9ea1a
                               at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/fmt/mod.rs:1153:17
   6:     0x560b5ef78def - std::io::Write::write_fmt::h0e1226b2b8d973fe
                               at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/io/mod.rs:1843:15
   7:     0x560b5ef7b734 - std::sys_common::backtrace::_print::hd2df4a083f6e69b8
                               at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x560b5ef7b734 - std::sys_common::backtrace::print::he907f6ad7eee41cb
                               at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x560b5ef7cd4b - std::panicking::default_hook::{{closure}}::h3926193b61c9ca9b
  10:     0x560b5ef7caa3 - std::panicking::default_hook::h25ba2457dea68e65
                               at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/panicking.rs:292:9
  11:     0x560b5ef7d1ed - std::panicking::rust_panic_with_hook::h0ad14d90dcf5224f
                               at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/panicking.rs:779:13
  12:     0x560b5ef7d089 - std::panicking::begin_panic_handler::{{closure}}::h4a1838a06f542647
                               at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/panicking.rs:649:13
  13:     0x560b5ef7be36 - std::sys_common::backtrace::__rust_end_short_backtrace::h77cc4dc3567ca904
                               at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/sys_common/backtrace.rs:171:18
  14:     0x560b5ef7cdf4 - rust_begin_unwind
                               at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/std/src/panicking.rs:645:5
  15:     0x560b5e884ac5 - core::panicking::panic_fmt::h940d4fd01a4b4fd1
                               at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/panicking.rs:72:14
  16:     0x560b5e884b83 - core::panicking::panic::h8ddd58dc57c2dc00
                               at /rustc/9b00956e56009bab2aa15d7bff10916599e3d6d6/library/core/src/panicking.rs:145:5
  17:     0x560b5e90dc7f - medians::algos::evenmedianu64::h1f0b2bc9e41ad162
                               at /home/marcalj/.cargo/registry/src/index.crates.io-6f17d22bba15001f/medians-3.0.11/src/algos.rs:383:23
  18:     0x560b5e90b717 - medians::medianu64::h044f3bca1d134ad4
                               at /home/marcalj/.cargo/registry/src/index.crates.io-6f17d22bba15001f/medians-3.0.11/src/lib.rs:86:34
```

Issue in:
https://docs.rs/medians/3.0.11/src/medians/algos.rs.html#383

**Comment by liborty** [comment]
Many thanks for reporting! I keep making this silly bug (subtracting from usize).
Fixed in version 3.0.12, just released.

**Comment by marcalj** [comment]
Thanks! I'm wondering why it's returning a reference [here](https://docs.rs/medians/3.0.12/src/medians/lib.rs.html#83-84), and not [here](https://docs.rs/medians/3.0.12/src/medians/lib.rs.html#88-90), since `oddmedianu64` and `evenmedianu64` doesn't return a reference.

Code for v3.0.11 always returning owned values: https://docs.rs/medians/3.0.11/src/medians/lib.rs.html#81-82

Seems that you released the last commit for v3.1 in this minor release, which also changes the return type of the median functions. I would suggest having a feature based branch in git for releases.

**Comment by liborty** [comment]
3.1 is on hold for now until I complete `evenmedu64` which is going to be a faster replacement of `evenmedianu64`. For that reason I have not released 3.1 on crates.io, yet (the primary repository). I guess I should not have made that early commit here either. For now please use 3.0.12 as is, from either. It should be in a consistent and working state.
